src/workerd/io/worker.c++

```diff
@@ -448,37 +448,106 @@ public:
       // We're on a request-serving thread.
       auto& ioContext = IoContext::current();
       timePoint = ioContext.now();
-    } else KJ_IF_MAYBE(info, inspectorTimerInfo) {
-      if (info->threadId == getCurrentThreadId()) {
-        // We're on an inspector-serving thread.
+    } else {
+      auto lockedState = state.lockExclusive();
+      KJ_IF_MAYBE(info, lockedState->inspectorTimerInfo) {
         timePoint = info->timer.now() + info->timerOffset
                   - kj::origin<kj::TimePoint>() + kj::UNIX_EPOCH;
+      } else {
+        // We're at script startup time -- just return the Epoch.
       }
     }
+    return (timePoint - kj::UNIX_EPOCH) / kj::MILLISECONDS;
+  }
 
-    // If we're on neither a request- nor inspector-serving thread, then we're at script startup
-    // time -- just return the Epoch.
+  void setInspectorTimerInfo(kj::Timer& timer, kj::Duration timerOffset) {
+    auto lockedState = state.lockExclusive();
+    lockedState->inspectorTimerInfo = InspectorTimerInfo { timer, timerOffset, getCurrentThreadId() };
+  }
 
-    return (timePoint - kj::UNIX_EPOCH) / kj::MILLISECONDS;
+  void setChannel(Worker::Isolate::InspectorChannelImpl& channel) {
+    auto lockedState = state.lockExclusive();
+    lockedState->channel = channel;
   }
 
-  // Nothing else. We ignore everything the inspector tells us, because we only care about the
-  // devtools inspector protocol, which is handled separately.
+  void resetChannel() {
+    auto lockedState = state.lockExclusive();
+    lockedState->channel = {};
+  }
 
-  void setInspectorTimerInfo(kj::Timer& timer, kj::Duration timerOffset) {
-    // Helper for attachInspector().
-    inspectorTimerInfo = InspectorTimerInfo { timer, timerOffset, getCurrentThreadId() };
+  void runMessageLoopOnPause(int contextGroupId) override {
+    // This method is called by v8 when a breakpoint or debugger statement is hit. This method
+    // processes debugger messages until `Debugger.resume()` is called, when v8 then calls
+    // `quitMessageLoopOnPause()`.
+    //
+    //NB This method is ultimately called from the `InspectorChannelImpl` and the isolate lock is
+    // held when this method is called.
+    if (isMultiTenantProcess()) {
+      // TODO: breakpoints are not supported for the multi-tenant case.
```

Comment:
Is this really a `TODO`?

Reply:
Yes, I'm afraid so. We continue to pass ew tests with this change, but I don't know enough about multi-tenant to know how well this would work at this point.

Specific gaps are:

1) For breakpoints you really need source maps to get started. That's easy to cons-up for local testing (add "//# sourceURL=file:///..."), but for ew it may require additional plumbing, e.g. fetch network resource.

2) What the expectation should be if with multiple requests and breakpoints / debugger statements. Naively, I assume we may end up with multiple isolates for the same scripts running on a single machine.

3) How to deploy a custom build of ew and test the ew preview case and test connecting CDT? https://wiki.cfdata.org/display/EW/How+to+debug+a+custom+edgeworker+build looks like a reasonable starting point.

Reply:
Dropped. I was really looking for a way to distinguish product from workerd since I would not consider this production ready. Probably better to just grok the process name.